File: src/java/magma/app/compile/value/Invokable.java
package magma.app.compile.value;

import magma.api.collect.Joiner;
import magma.api.collect.list.Iterable;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.app.ValueCompiler;
import magma.app.compile.CompileState;
import magma.app.compile.node.Node;
import magma.app.compile.type.PrimitiveType;
import magma.app.compile.type.Type;

public record Invokable(Node node, Iterable<Value> args) implements Value {
    public String generate() {
        var joinedArguments = this.joinArgs();
        return ValueCompiler.getString(this.node) + "(" + joinedArguments + ")";
    }

    public String joinArgs() {
        return this.args.iter()
                .map((Value value) -> {
                    return ValueCompiler.generateValue(value);
                })
                .collect(new Joiner(", "))
                .orElse("");
    }

    public Option<Value> toValue() {
        return new Some<Value>(this);
    }

    public Type resolve(CompileState state) {
        return PrimitiveType.Unknown;
    }

}

File: src/ts/magma/app/compile/value/Invokable.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Iterable } from "../../../../magma/api/collect/list/Iterable";
import { ValueCompiler } from "../../../../magma/app/ValueCompiler";
import { Joiner } from "../../../../magma/api/collect/Joiner";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { Type } from "../../../../magma/app/compile/type/Type";
import { CompileState } from "../../../../magma/app/compile/CompileState";
import { PrimitiveType } from "../../../../magma/app/compile/type/PrimitiveType";
export class Invokable implements Node {
	node: Node;
	args: Iterable<Node>;
	constructor (node: Node, args: Iterable<Node>) {
		this.node = node;
		this.args = args;
	}
	generate(): string {
		let joinedArguments = this.joinArgs()/*unknown*/;
		return ValueCompiler.getString(this.node) + "(" + joinedArguments + ")"/*unknown*/;
	}
	joinArgs(): string {
		return this.args.iter().map((value: Node) => {
			return ValueCompiler.generateValue(value)/*unknown*/;
		}).collect(new Joiner(", ")).orElse("")/*unknown*/;
	}
	toNode(): Option<Node> {
		return new Some<Node>(this)/*unknown*/;
	}
	resolve(state: CompileState): Type {
		return PrimitiveType.Unknown/*unknown*/;
	}
}
